### PR TITLE
Fix mix compile error when there is a wrong project path.

### PR DESCRIPTION
**Pull request by soteras**
This PR aims to solve the issue #7531, the problem is Mix.Task.run("compile.all", args) returns a empty list when project path is wrong. 

```
def project do
    [
      app: :apps_path_bug,
      apps_path: "this_path_does_not_exist_or_is_empty"
    ]
 end
```

### CODE REVIEW

File: lib/mix/lib/mix/tasks/compile.ex [hunk 1]
```diff
@@ -91,10 +91,16 @@ defmodule Mix.Tasks.Compile do
     load_erl_config(opts)
 
     {res, diagnostics} =
-      Mix.Task.run("compile.all", args)
-      |> List.wrap()
-      |> Enum.map(&Mix.Task.Compiler.normalize(&1, :all))
-      |> Enum.reduce(&merge_diagnostics/2)
```
Comment (josevalim) [inline lib/mix/lib/mix/tasks/compile.ex]: The simplest fix here would be to replace `Enum.reduce(&merge_diagnostics/2)` by `Enum.reduce({:ok, []}, &merge_diagnostics/2)`. Then we just ensure that calling the umbrella app does not raise.

Comment (soteras) [reply lib/mix/lib/mix/tasks/compile.ex]: Something like that?

```
diagnostics =
      Mix.Task.run("compile.all", args) |> List.wrap()
      |> Enum.map(&Mix.Task.Compiler.normalize(&1, :all))

    {res, diagnostics} =
      case diagnostics do
        [] ->
          Enum.reduce(diagnostics, {:error, []}, &merge_diagnostics/2)

        _ ->
          Enum.reduce(diagnostics, &merge_diagnostics/2)
      end
```

File: lib/mix/test/mix/tasks/compile_test.exs [hunk 2]
```diff
@@ -9,6 +9,12 @@ defmodule Mix.Tasks.CompileTest do
     end
   end
 
+  defmodule WrongPath do
+    def project do
+      [app: :apps_path_bug, apps_path: "this_path_does_not_exist_or_is_empty"]
```
Comment (eksperimental) [inline lib/mix/test/mix/tasks/compile_test.exs]: It's wrong the part of the atom that says its_empty. Maybe another module with empty path can be created and tested

File: lib/mix/lib/mix/tasks/compile.ex [hunk 3]
```diff
@@ -90,11 +90,19 @@ defmodule Mix.Tasks.Compile do
...
+    diagnostics =
       Mix.Task.run("compile.all", args)
       |> List.wrap()
       |> Enum.map(&Mix.Task.Compiler.normalize(&1, :all))
-      |> Enum.reduce(&merge_diagnostics/2)
+
+    {res, diagnostics} =
+      case diagnostics do
+        [] ->
+          Enum.reduce(diagnostics, {:ok, []}, &merge_diagnostics/2)
```
Comment (josevalim) [inline lib/mix/lib/mix/tasks/compile.ex]: We can keep the code exactly the same as before, we just need to change one line where we replace Enum.reduce/2 by a Enum.reduce/3. :)

Comment (soteras) [reply lib/mix/lib/mix/tasks/compile.ex]: I need to change old tests. Do you agree?
Assertion with == failed
     code:  assert Mix.Task.run("compile", ["--verbose"]) == {:noop, []}
     left:  {:ok, []}
     right: {:noop, []}
     stacktrace:
       lib/mix/test/mix/tasks/compile_test.exs:54: anonymous fn/0 in Mix.Tasks.CompileTest."test compile a project with mixfile"/1
       (elixir) lib/file.ex:1428: File.cd!/2
       lib/mix/test/test_helper.exs:114: MixTest.Case.in_fixture/3
       lib/mix/test/mix/tasks/compile_test.exs:44: (test)

Comment (soteras) [reply lib/mix/lib/mix/tasks/compile.ex]: Maybe that:

Enum.reduce({:noop, []}, &merge_diagnostics/2)

File: lib/mix/test/mix/tasks/compile_test.exs (hunk 2 above)
Comment (soteras) [reply lib/mix/test/mix/tasks/compile_test.exs]: Yes i'll create

File: lib/mix/lib/mix/tasks/compile.ex (hunk 3 above)
Comment (josevalim) [reply lib/mix/lib/mix/tasks/compile.ex]: Oh, I see. Yes, `|> Enum.reduce({:noop, []}, &merge_diagnostics/2)` should work!

File: lib/mix/test/mix/tasks/compile_test.exs [hunk 4]
```diff
@@ -157,4 +169,20 @@ defmodule Mix.Tasks.CompileTest do
   after
     Application.delete_env(:erl_config_app, :value)
   end
+
+  test "compile a project with wrong path" do
```
Comment (eksperimental) [inline lib/mix/test/mix/tasks/compile_test.exs]: Tests should be in present tense, not in imperative. So it should read:
"compiles a project"

File: lib/mix/test/mix/tasks/compile_test.exs [hunk 5]
```diff
@@ -157,4 +169,20 @@ defmodule Mix.Tasks.CompileTest do
...
+    Mix.Project.push(WrongPath)
+
+    ExUnit.CaptureIO.capture_io(fn ->
+      assert {:noop, []} = Mix.Task.run("compile", ["--force", "--return-errors"])
+    end)
+  end
+
+  test "compile a project with empty path" do
```
Comment (eksperimental) [inline lib/mix/test/mix/tasks/compile_test.exs]: it should read:
"compiles a project"

File: lib/mix/test/mix/tasks/compile_test.exs (hunk 4 above)
Comment (soteras) [reply lib/mix/test/mix/tasks/compile_test.exs]: English is not my first language but if you open the file lib/mix/test/mix/tasks/compile_test.exs there're a lot tests with "compile" they are wrong too?

Comment (eksperimental) [reply lib/mix/test/mix/tasks/compile_test.exs]: Yes, they should be corrected.
I think it was myself and/or @lexmag who standardized all the tenses but those that you see have been introduced relatively recently.
You can see it has been done before in c6f83b5 for example

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request merged**